sup_doc/dte/src/doc/user.gd/book-format.1
...\" Copyright 1991,1992,1993 Open Software Foundation, Inc.,
...\" Cambridge, Massachusetts
...\" All rights reserved.
...\"
...\" @OSF_FREE_COPYRIGHT@
...\"
...\" HISTORY
...\" $Log: book-format.1,v $
...\" Revision 1.1.2.7  1994/06/24  15:16:08  fred
...\" 	free copyright
...\" 	[1994/06/23  20:15:18  fred]
...\"
...\" Revision 1.1.2.6  1994/06/23  18:40:04  fred
...\" 	free copyright
...\" 	[1994/06/22  19:34:10  fred]
...\" 
...\" Revision 1.1.2.5  1993/12/09  14:21:55  rcb
...\" 	Made changes requested by DTE CR 449, including adding -sml and -nosml to
...\" 	SYNOPSIS, changing order of flag descriptions to match SYNOPSIS, and
...\" 	correcting a typo.
...\" 	[1993/12/09  14:21:45  rcb]
...\" 
...\" Revision 1.1.2.4  1993/04/10  01:22:40  bowe
...\" 	Add OSF copyright.
...\" 	[1993/04/10  01:20:22  bowe]
...\" 
...\" Revision 1.1.2.3  1993/02/26  13:58:06  bowe
...\" 	Documented the "-nochange" option.
...\" 	[1993/02/26  13:52:06  bowe]
...\" 
...\" Revision 1.1.2.2  1992/07/07  22:26:28  bowe
...\" 	Initial rev.
...\" 	[1992/07/07  22:24:31  bowe]
...\" 
...\" $EndLog$
...\" $Header: /u1/rcs/dte/doc/user.gd/book-format.1,v 1.1.2.7 1994/06/24 15:16:08 fred Exp $
...\"
.TH book-format 1
.SH NAME
book-format \- format book source files
.SH SYNOPSIS
.B book-format
[ \fB-n\fP | \fB-t\fP | \fB-x\fP ] 
[ \fB-sml\fP | \fB-nosml\fP ]
[ \fB-d\fP | \fB-p\fP ] [ \fB-v\fP ]
[ \fB-T\fItypesetter\fP ] [ \fB-D\fIsymbol\fP ...] [ \fB-showifdef\fP ]
[ \fB-des\fP \fIfile\fP ]
[ \fB-libdir\fP \fIdirectory\fP ] [ \fB-output\fP \fIdirectory\fP ]
[ \fB-mac\fP \fImacro-file\fP ]
[ \fB-book\fP \fIname\fP ]
[ \fB-alt\fP \fIdirectory\fP ]
[ \fB-version\fP ]
[ \fB-printorder\fP ]
[ \fB-nochange\fP ]
[ \fB-help\fP ]
[ \fBbody\fP ] [ \fBtoc\fP ] [ \fBindex\fP ] [ \fBglossary\fP ]
[ \fIpiece\fP ] [ \fBall\fP ]
.SH DESCRIPTION
The
.B book-format
command is used to format books or pieces of books.
Options may appear in any order, except the pieces to format,
which must appear last.
.PP
A book is described by a \fIdescription(4)\fP file which lists all the
pieces of the book in order.
Pieces include the title page, preface, and appendices.
There is one piece per file; separation is strictly up to the writer.
The description file may have comments, lines whose first character is `#',
and blank lines.
All other lines are significant.
.PP
Output is placed in subdirectories, one formatted piece per file.
While building a book, intermediate information, that is state information
at the end of each piece, is kept in a directory called \fISync\fP.
This information is used when building a single piece of the book.
.PP
\fBbook-format\fP calls the program \fBformat-piece\fP,
which does all the work, calling nroff or troff along with the appropriate
preprocessors and macros.
.SH "OPTIONS"
The following are the possible parameters to \fBbook-format\fP:
.TP 8
.B -n
Build with nroff.
.TP 8
.B -t
Build with troff (default).
.TP 8
.B -x
Build with express.
.TP 8
.B -sml
Use the SML macros instead of plain -mm.
.TP 8
.B -nosml
Do not use the SML macros instead of plain -mm (default).
.TP 8
.B -d
Draft formatting.
.TP 8
.B -p
Publish formatting (default).
.TP 8
.B -v
Displays some debugging information that provides more detail about
the processing steps, including the sequence of the macro packages called.
.TP 8
.B -T\fItypesetter\fP
if using troff, use \fItypesetter\fP (default \fIps\fP).
.TP 8
.B -D\fIsymbol\fP
Pipe document source through \fBrmifdef\fP, defining \fIsymbol\fP.
This will only happen if the directive \fI...\e"build rmifdef\fP
is within an initial, contiguous block of comments in the source files.
You can have multiple -D options on the command line.
-D options may be included in the description file in place of the command
line.
If symbols are defined on the command line in this way, all -D directives in
the description file are ignored.
-D with no \fIsymbol\fP following will specify that the files are to be
processed through \fBrmifdef\fP with no symbols defined.
.TP 8
.B -showifdef
Show all #if, #ifdef, #else, and #endif directives by making them bold and
surrounding them by a small amount of vertical space.
This option disables all #ifdef processing with rmifdef.
.TP 8
.B -des \fIfile\fP
specify alternate description file (default \fI./Description\fP).
.TP 8
.B -libdir \fIdirectory\fP
Set library directory of Documentation Tools Environment (DTE)
(default $DTE_LIBDIR, if set, or /project/tools/libdata).
.TP 8
.B -output \fIdirectory\fP
Set directory for output (default $OUTPUT, if set, or \fI./Output\fP).
.TP 8
.B -mac \fImacro-file\fP
Include \fImacro-file\fP after other macro files and before your source file.
To include several files either use multiple \fB-mac\fP \fImacro-file\fP
arguments or use quotes around multiple words for \fImacro-file\fP.
.TP 8
.B -book \fIname\fP
When creating a glossary, retrieve only those definitions from the
glossary database associated with book \fIname\fP, where \fIname\fP
must be a book identifier listed in the \fBbookdefs.mac\fP file.
If this option is not specified, definitions for all books will be retrieved.
The book name may also be specified in the description file.
A book name specified on the command line overrides that specified in
the description file.
.TP 8
.B -alt \fIdirectory\fP
When building a book some source files may be located in \fIdirectory\fP,
allowing writers to keep a subset of files of a book in a working directory
during development of that book.
.TP 8
.B -version
Display version information about \fBbook-format\fP, then exit.
This is useful for reporting bugs to the maintainers.
.TP 8
.B -printorder
Name the output files so that they will list in the order of the pieces in
the description file.
Names are preceded by a 3-digit number related to the piece number.
This allows one to print the files in order with \fIlpr *\fP.
.TP 8
.B -nochange
Do not produce change markers in the output.
(These are produced by the zA and zZ macros.)
.TP 8
.B -help
Print a help message summarizing options and usage.
.TP 8
.B body
Builds the entire document body.
.TP 8
.B toc
Builds the document table of contents.
.TP 8
.B index
Builds the document index.
.TP 8
.B glossary
Builds the document glossary.
.TP 8
.I piece
Builds only that named piece
(piece names are the first column in the description file).
.TP 8
.B all
Build everything (all pieces, toc, index, glossary.
.PP
By default \fBbook-format\fP looks for the description file in the
current directory.
.PP
Formatting \fBall\fP creates synchronization files that are used later if
the user wants to format a \fBpiece\fP of the book.
This information is also needed for table of contents, index, and
glossary generation.
Individual pieces can not be formatted unless \fBbook-format ... all\fP
has been done.
.SH "FILES"
.TP 16
$DTE_LIBDIR/macros
troff/nroff macros directory
.TP 16
$DTE_LIBDIR/glossary
glossary database directory
.TP 16
\&./Description
description file describing book
.SH "CUSTOMIZATION"
.PP
The DTE assumes that Elan's Documenter's Workbench (DWB) programs are available.
If another vendor's DWB is installed, or if alternate programs are desired,
environment variables may be set to define which programs will be used for the
various DWB components.
This feature is also useful if you want to pass special arguments to
these programs.
Below is a list of variables and the default values.
.sp
.nf
.ft CW
.in +1cm
TROFF          troff
NROFF          nroff
TYPESETTER     ps
DTE_OUTFILTER  eps
RMIFDEF        rmifdef
EQN            eqn
NEQN           neqn
PIC            pic
TBL            tbl
.in 0
.fi
.ft P
.PP
For example, to format for a Hewlett-Packard LaserJet printer, set the
environment variable TYPESETTER to \fIhplj\fP and DTE_OUTFILTER to \fIt2hplj\fP.
This will run \fItroff\fP with the flag -T\fIhplj\fP and pipe the
output though the program \fIt2hplj\fP.
\fItroff\fP output will be placed in the directory named Output/.../\fIhplj\fP.
.P
These environment variables may also be set in a configuration file
in a user's home directory called \fB.dterc\fP.
The syntax is that of the Bourne shell.
The variables need not be exported in the config file.

.SH "BUGS"
.PP
This command is intended for formatting entire books.
It would be nice to have commands for shorter works, like technical papers.
.PP
When trying to format a single piece of a book, glossary, table-of-contents,
or index, \fBbook-format\fP is not robust enough to recover from the user's
failure to do have previously done
\fBbook-format ... all\fP.
(\fBbook-format ... all\fP produces synchronization information about the
state between pieces.)
\fBbook-format\fP will exit ungracefully.
.SH "SEE ALSO"
dmm(1), dman(1), description(4), glossary(4)
.br
\fIDTE User's Guide and Reference\fP


sup_doc/dte/src/book-format/book-format.sh
#!/bin/sh
#
# Copyright 1991,1992,1993 Open Software Foundation, Inc.,
# Cambridge, Massachusetts
# All rights reserved.
#
#
# @OSF_FREE_COPYRIGHT@
# 
# HISTORY
# $Log: book-format.sh,v $
# Revision 1.1.2.7  1994/06/23  19:48:22  fred
# 	free copyright
# 	[1994/06/23  19:39:24  fred]
#
# Revision 1.1.2.6  1994/06/23  18:37:15  fred
# 	free copyright
# 	[1994/06/22  15:26:06  fred]
# 
# Revision 1.1.2.5  1993/04/10  01:02:56  bowe
# 	Add OSF copyright.
# 	[1993/04/10  00:59:57  bowe]
# 
# Revision 1.1.2.4  1993/02/26  13:58:03  bowe
# 	Added new feature, -nochange, which will eliminated change markers on
# 	output (zA, zZ macros).  Done by creating a tmp macro file to remove zA
# 	and zZ and reading it as if user had done "-mac that-tmp-file" option.
# 	[1993/02/26  13:57:33  bowe]
# 
# Revision 1.1.2.3  1992/12/10  18:37:44  fred
# 	various changes to implement the new master index
# 	[1992/12/10  18:35:40  fred]
# 
# Revision 1.1.2.2  1992/07/07  14:51:00  bowe
# 	Initial rev.
# 	[1992/07/07  14:48:02  bowe]
# 
# $EndLog$
#
# $Header: /u1/rcs/dte/book-format/book-format.sh,v 1.1.2.7 1994/06/23 19:48:22 fred Exp $
#
#  book build
#
#  High-level control program for building books.  Simply figures out what
#  to do and calls other programs.
#

Version="\$Revision: 1.1.2.7 $, \$Date: 1994/06/23 19:48:22 $";

DEFAULT_DTE_LIBDIR=/project/tools/libdata

if [ -z "$DTE_LIBDIR" ]	; then DTE_LIBDIR=$DEFAULT_DTE_LIBDIR ; fi
if [ -z "$OUTPUT" ]	; then OUTPUT=Output ; fi
if [ -z "$FMT_PIECE" ]	; then FMT_PIECE="fmt-piece" ; fi
if [ -z "$SYNC_DIR" ]	; then SYNC_DIR="Sync" ; fi

doNroff=false
doTroff=false
doExpress=false
formatType="d"
Form=draft
subdir1=draft
Language=PostScript
typeset=""
verbose=false

tmpSync=false
doBody=false
doToc=false
doIndex=false
doMIndex=false
doGloss=false
doAll=false
glossArgs=""
bookName=""
doSML=""
SMLspecified=false
FMTspecified=false
ROFFspecified=false
defines=""
formatOption=-t

descFile=Description
MIdescFile=MI.Description
Piece=""
doNothing=true
extraArgs=""
doPrintOrder=false

# look at personal defaults
if [ -f $HOME/.dterc ] ; then
    . $HOME/.dterc
    export DTE_LIBDIR DTE_OUTFILTER TYPESETTER
    export TROFF NROFF TBL PIC NEQN EQN RMIFDEF
fi

if [ -n "$DTE_TMPSYNC" ] ; then tmpSync=true ; fi

# extra macros generated by this program
ext_mac=/tmp/ext_mac$$
trap "rm -rf $ext_mac ; exit $?" 0 1 2 3 15

# parse command line

while [ $# -gt 0 ] ; do
    case "$1" in

	-n)	Language=ASCII
		formatOption=-n
		doNroff=true
		ROFFspecified=true
		subdir2=ascii
		;;
	-t)	Language=PostScript
		formatOption=-t
		doNroff=false
		doTroff=true
		ROFFspecified=true
		;;
	-x)	Language=Express
		formatOption=-x
		doNroff=false
		doExpress=true
		ROFFspecified=true
		subdir2=dito
		;;
	-draft | -d)
		Form=draft
		formatType=d
		FMTspecified=true
		subdir1=draft
		;;
	-pub* | -p)
		Form=pub
		formatType=p
		FMTspecified=true
		subdir1=publish
		;;
	-nosml)	doSML=""
		SMLspecified=true
		;;
	-sml)	doSML="-sml"
		SMLspecified=true
		;;
	-T*)	typeset=$1
		TYPESETTER=`echo $1 | sed "s/-T//"`
		;;
	-libdir | -root)
		shift
		DTE_LIBDIR=$1
		;;
	-out*)	shift
		OUTPUT_DIR=$1
		;;
	-alt)	shift
		extraArgs="$extraArgs -alt $1"
		;;
	-des*)	shift
		descFile=$1
		extraArgs="$extraArgs -des $descFile"
		;;
	-book)	shift
		bookName=$1
		;;
	-D*)	defines="$defines $1" ;;
	-U*)	defines="$defines $1" ;;
	-showif*)
		extraArgs="$extraArgs -showifdef"
		;;
	-v)	verbose=true
		extraArgs="$extraArgs -v"
		;;
	-mac)	extraArgs="$extraArgs -mac \"$2\""
		shift
		;;
	-tmpsync)
		tmpSync=true
		;;
	body)	doNothing=false
		doBody=true
		;;
	toc)	doNothing=false
		doToc=true
		;;
	index)	doNothing=false
		doIndex=true
		;;
	master-index)
		doNothing=false
		doMIndex=true
		;;
	gloss*)	doNothing=false
		doGloss=true
		;;
	all | -all)
		doNothing=false
		doBody=true
		doToc=true
		doIndex=true
		doGloss=true
		doAll=true
		;;
	-version)
		echo "`basename $0`: $Version"
		exit 0
		;;
	-printorder)
		doPrintOrder=true
		extraArgs="$extraArgs -printorder"
		;;
	-nochange*)
		extraArgs="$extraArgs -mac $ext_mac"
		touch $ext_mac
		echo ".rm zA zZ" >> $ext_mac
		;;
	-h*)	doNothing=true
		break
		;;
	*)	doNothing=false
		Piece="$Piece $1"
		;;
    esac
    shift
done

# ________________________________________________________________________

if `$doNothing`; then
    cat << E_O_F

synopsis:  book-format [-n|-t|-x] [-d|-p] [-Ttypesetter] [-des <file>]
	[-libdir <dir>] [-output <dir>] [-D<symbol> ...] [-showifdef]
	[body] [toc] [index] [gloss] [piece ...] [all]

    -t		Build with troff (default)
    -n		Build with nroff
    -x		Build with express
    -sml	Build with sml macros
    -nosml	Build without sml macros (default)
    -draft|-d	Draft formatting (default)
    -pub|-p	Publish formatting
    -mac <macs>	include macros <macs> after others.
    -T<typesetter>
		Use the named typesetter (don't postprocess) (default "ps")

    -out <dir>	Set directory for output to <dir>
    -des <file>	Specify alternate Description file to <file>
    -alt <dir>	Search directory <dir> for files not found in local directory
    -libdir <dir> Set root directory of DTE to <dir>
    -book <book>  Specify name of book to screen by when formatting glossary

    -D<sybmol>	Process #ifdef/#else/#endif (cpp-like), defining <symbol>.
    -showifdef	Embolden #ifdef/#else/#endif in the output.  Turns off other
		#ifdef processing.

    -printorder	Name the output files so that 'lpr *' will print them in order

    -nochange	Omit change markers on output

    body	Builds the document body
    toc		Builds the document table of contents
    index	Builds the document index
    gloss	Builds the document glossary
    piece	Builds only the named pieces (names from Description file)
    all		Builds everything (body, toc, glossary, index)

E_O_F
    exit 1
fi

# ________________________________________________________________________

# working locations and format control places

if [ ! -d "$DTE_LIBDIR" ] ; then
    echo "DTE_LIBDIR is set to $DTE_LIBDIR, a non-existent directory."
    exit 1
fi

if $doMIndex ; then
    if [ ! -r $MIdescFile ] ; then
	echo "Can't find Master Index Description file ($MIdescFile)"
	exit 1
    fi
elif [ ! -r $descFile ] ; then
    echo "Can't find book Description file ($descFile)"
    exit 1
fi

PATH=/usr/local/bin:$PATH
export DTE_LIBDIR PATH

if $doMIndex ; then
    descFile=$MIdescFile
else
    if $tmpSync ; then
	SYNC_DIR=/tmp/Sync+$$
	trap "rm -rf $SYNC_DIR $ext_mac ; exit $?" 0 1 2 3 15
    fi
    if [ ! -d $SYNC_DIR ] ; then
	mkdir $SYNC_DIR
	if [ ! -d $SYNC_DIR ] ; then
	    echo "Can not create sync directory ($SYNC_DIR)."
	    exit 1
	fi
    fi
    export SYNC_DIR
fi

# ________________________________________________________________________

# look thru the description file for extra instructions

if [ -z "$bookName" ] ; then
    bookName="`egrep '^##.*BOOK' $descFile | head -1 | sed -e 's/^.*BOOK//'`"
fi

if [ -z "$defines" ] ; then
    defines="`egrep '^##.*DEFINE' $descFile | sed -e 's/^.*DEFINE//'`"
fi

if [ -n "$bookName" ] ; then glossArgs="-book $bookName" ;fi

if [ -n "$defines" ] ; then extraArgs="$extraArgs $defines" ; fi

#  These can be used if writer NEVER wants a glossary, toc, index when
#  doing "book-format all".  (saves processing time).
if egrep '^##.*NO GLOSS' $descFile >/dev/null ; then
    doGloss=false
    if $verbose ; then echo "Will skip glossary. (per $descFile)" ; fi
fi
if egrep '^##.*NO TOC' $descFile >/dev/null ; then
    doToc=false
    if $verbose ; then echo "Will skip toc. (per $descFile)" ; fi
fi
if egrep '^##.*NO INDEX' $descFile >/dev/null ; then
    doIndex=false
    if $verbose ; then echo "Will skip index. (per $descFile)" ; fi
fi

if `$SMLspecified` ; then
    true
else
    if egrep '^##[ 	]*NO *SML' $descFile >/dev/null ; then
	doSML=""
	a=" NOT"
    elif egrep '^##[ 	]*SML' $descFile >/dev/null ; then
	doSML="-sml"
	a=""
    fi
    if $verbose ; then echo "Will$a use SML macros. (per $descFile)" ; fi
fi

if $FMTspecified ; then
    true
else
    if egrep '^##[ 	]*DRAFT' $descFile >/dev/null ; then
	Form=draft ; formatType=d
    elif egrep '^##[ 	]*PUBLISH' $descFile >/dev/null ; then
	Form=pub   ; formatType=p
    fi
    if $verbose ; then
	echo "Formatting with '$Form' format. (per $descFile)"
    fi
fi

if $ROFFspecified ; then
    true
else
    if egrep '^##[ 	]*TROFF' $descFile >/dev/null ; then
	Language=PostScript ; formatOption=-t ; doNroff=false ; doTroff=true
    elif egrep '^##[ 	]*NROFF' $descFile >/dev/null ; then
	Language=ASCII      ; formatOption=-n ; doNroff=true
    elif egrep '^##[ 	]*EXPRESS' $descFile >/dev/null ; then
	Language=Express    ; formatOption=-x ; doNroff=false ; doExpress=true
    fi
    if $verbose ; then
	echo "Using '$Language' and '$formatOption'. (per $descFile)"
    fi
fi

if egrep '^##.*PRINTORDER' $descFile >/dev/null ; then
    doPrintOrder=true
    extraArgs="$extraArgs -printorder"
    if $verbose ; then echo "Using print ordering (per $descFile)" ; fi
fi

# ________________________________________________________________________

if [ -z "$subdir2" ] ; then
    if [ -n "$TYPESETTER" ] ; then subdir2=$TYPESETTER
    else subdir2="ps"			: default typesetter...
    fi
fi

if [ -z "$OUTPUT_DIR" ] ; then
    OUTPUT_DIR=$OUTPUT/$subdir1/$subdir2
    if [ ! -d $OUTPUT ] ; then mkdir $OUTPUT ; fi
    if [ ! -d $OUTPUT/$subdir1 ] ; then mkdir $OUTPUT/$subdir1 ; fi
    if [ ! -d $OUTPUT_DIR ] ; then mkdir $OUTPUT_DIR ; fi
else
    if [ ! -d $OUTPUT_DIR ] ; then mkdir $OUTPUT_DIR ; fi
fi
export OUTPUT_DIR

# ________________________________________________________________________
#
#	the body stuff

#  If doing "all", remove sync files so that if writer added/removed a
#  file, things won't get screwed up.  A safty catch.
if $doAll ; then rm -f $SYNC_DIR/$formatType$formatOption.* ; fi

echo ""
echo "Doing '$Form' formatting into $Language"

#  See if requested to do entire body.

if $doBody; then
    count=1
    maxcount=`sed -e '/^#/d' -e '/^$/d' $descFile | wc -l`

    while [ $count -le $maxcount ] ; do
	if $verbose ; then
	    echo "_______________________________"
	    date
	    echo args to ${FMT_PIECE}: $formatOption -$formatType $doSML \
			$extraArgs $count $nroffOptions $typeset
	fi
	${FMT_PIECE} $formatOption -$formatType $doSML $extraArgs $count \
			"$nroffOptions" $typeset
	count=`expr $count + 1`
    done
fi

#  See if requested to do a single piece.

if [ "$Piece" != "" ]; then
    for P in $Piece ; do
	one='$1'	: HACK
	PieceNum=`sed -e '/^#/d' -e '/^$/d' $descFile | \
		    awk "$one ~ /^${P}$/ {print NR}"`
	if [ `echo $PieceNum | wc -w` -gt 1 ] ; then
	    echo "There is more than one piece with the same name, $Piece,"
	    echo "in the Description file ($descFile):"
	    grep $Piece $descFile
	    exit 1
	fi
	if [ "$PieceNum" -gt 0 -a "$PieceNum" -lt 5000 ]; then
	    if $verbose ; then
		echo args to ${FMT_PIECE}: $formatOption -$formatType $doSML\
			$extraArgs $PieceNum $nroffOptions $typeset
	    fi
	    ${FMT_PIECE} $formatOption -$formatType $doSML $extraArgs \
			$PieceNum "$nroffOptions" $typeset
	else
	    echo "can't find piece $Piece"
	    exit 2
	fi
    done
fi

# ________________________________________________________________________

# Do Table of Contents, Index, Glossary, if requested

if $doIndex && $atOSF; then
    fmt-index $formatOption -$formatType $typeset $extraArgs
fi

if $doGloss && $atOSF; then
    fmt-gloss $formatOption -$formatType $typeset $glossArgs $extraArgs
fi

if $doToc && $atOSF ; then
    fmt-toc $formatOption -$formatType $typeset $extraArgs
fi

if $doMIndex && $atOSF ; then
    fmt-index $formatOption -$formatType $typeset $extraArgs master
fi

if $verbose ; then date ; fi
echo ""

